website_order_invoice_automation/models/sale_order.py

```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018 Quartile Limited
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from odoo import models, fields
+
+_logger = logging.getLogger(__name__)
+
+class SaleOrder(models.Model):
+    _inherit = 'sale.order'
+
+    def _generate_and_validate_invoice(self):
+        ctx_company = {'company_id': self.company_id.id,
+                       'force_company': self.company_id.id}
+        created_invoice = self.with_context(
+            **ctx_company).action_invoice_create()
+        created_invoice = self.env['account.invoice'].browse(
+            created_invoice).with_context(**ctx_company)
+
+        if created_invoice:
+            _logger.info('Auto-generated invoice %s (ID %s) for %s (ID %s)',
+                         created_invoice.name, created_invoice.id,
+                         self.name, self.id)
+
+            created_invoice.action_invoice_open()
+        else:
+            _logger.warning('Could not auto-generate invoice for %s (ID %s)',
+                            self.name, self.id)
+
+    def _is_invoiceable(self):
+        invoiceable_order = False
+        for order_line in self.order_line:
+            # Check whether is invoiceable lines
+            if not invoiceable_order and order_line.qty_to_invoice > 0:
+                invoiceable_order = True
+            # Check whether the order is fully delivered, return False if
+            # any line is not fully delivered (except service product /
+            # shipping cost)
+            if (order_line._get_delivered_qty() < order_line.product_uom_qty) \
+                    and order_line.product_id.type != 'service' \
+                    and not order_line.product_id.is_shipping_cost:
+                return False
```

Comment:
Haven't fully understood this part of the logic.  Is this needed because we cannot rely on `order_id.qty_to_invoice` field for the products that are being delivered (i.e. the field is not updated by the time this method is run)?